### PR TITLE
repr,sql: handle set operations on numerics another way

### PR DESCRIPTION
**Pull request by benesch**
I'm not sure how much I like this, but it is maybe an option.

----

Rather than canonicalizing numeric data whenever we pack it into a Row,
instead only canonicalize numeric data before it heads into a set
operation, like UNION or DISTINCT. This means we preserve trailing zeros
whenever possible while still ensuring reasonable semantics for set
operations.

Note that we're still not quite PostgreSQL compatible on set operations.
When PostgreSQL encounters `DISTINCT ... VALUES (1.0), (1.00)`, it
promises to output `1.0` if you've specified an `ORDER BY`, because 1.0
appears first. But we'll output `1` because we always trim off trailing
zeros from all data that flows through a set operation.

Importantly, this approach should generalize reasonably well to the
`char` type, which has a similar issue around trailing whitespace.
`char` will be a bit harder, though, because we'll need to cast chars to
strings on the way in to set operations, and then cast them back to
chars on the way out.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/7572)

### CODE REVIEW

**Review by frankmcsherry — COMMENTED** [review]
I can't really tell what this PR means to do. It seems like it removes canonicalization of decimals, which will break joins on `dec1 = dec2`, make grouping by decimal keys distinguish between different non-canonical decimals, etc.

At the same time, it doesn't touch `Row::eq` nor optimization or dataflow, so I don't have a strong feeling if this is an incremental positive for the SQL team.

**Comment by benesch** [comment]
Oops, I failed to press enter on my last comment last night. Just posted it now: https://github.com/MaterializeInc/materialize/pull/7569#issuecomment-888262474.

File: src/sql/src/plan/query.rs [hunk 1]
```diff
@@ -977,6 +977,43 @@ fn plan_subquery(
...
+    restrict_to: Option<&[usize]>,
+) -> HirRelationExpr {
+    let should_consider = |i| match restrict_to {
+        Some(cols) => cols.contains(&i),
+        None => true,
+    };
+    let mut map_exprs = vec![];
+    let mut project_exprs = vec![];
+    for (i, ty) in types.iter().enumerate() {
+        if should_consider(i) && matches!(ty.scalar_type, ScalarType::Numeric { scale: None }) {
```
Comment (sploiselle) [inline src/sql/src/plan/query.rs:1001]: realize this is a prototype, but just want to note for myself:
- This probably wants to match `Numeric` irrespective of scale
- Needs to descend into `List`s to see if their elements are `Numeric`, which would also need canonicalization.

File: src/sql/src/func.rs [hunk 2]
```diff
@@ -2395,7 +2395,11 @@ lazy_static! {
                 params!(ArrayAny, ArrayAny) => BinaryFunc::Gte, 1075;
             },
             "=" => Scalar {
-                params!(Numeric, Numeric) => BinaryFunc::Eq, 1752;
+                params!(Numeric, Numeric) => Operation::binary(|_ecx, lhs, rhs| {
```
Comment (sploiselle) [inline src/sql/src/func.rs:2398]: one last general note that this change would need to get propagated to `List`s everywhere, as well

**Review by sploiselle — COMMENTED** [review]
Seems tractable, though the invariant between datum and row equality become subtle, right? i.e. rows are only guaranteed to express the same equality as their datums if you expressed an equality on the row, if I'm reading this correctly.

**Comment by benesch** [comment]
So I think we would actually want to make Datum and Row equality 1:1. So
Datum’s eq on decimals should do the dumb thing (compare all the fields),
to shake out any places where the SQL planner forgot to inject a
canonicalization call.

**Comment by sploiselle** [comment]
@benesch My point was poorly phrased--was trying to point to the fact that things will fail anywhere the planner forgets to inject a canonicalization call; sounds like we're on the same page.

**Comment by benesch** [comment]
Closing as stale.

**Pull request closed without merging**